src/Chapter3.hs

```diff
@@ -1007,6 +1157,147 @@ contestants, and write a function that decides the outcome of a fight!
+    , fknightRoundsBoostedDefense :: Int
+    }
+    deriving (Show)
+
+noAttack = MkAttack 0
+
+attackKnight :: Knight -> (Knight, Attack)
+attackKnight k = case k of
+    FightingKnight _ _ atk _ _  -> (k, atk)
+    Knight _ atk _              -> (k, atk)
+
+usePotionKnight :: Knight -> (Knight, Attack) 
+usePotionKnight (Knight  h atk def) = (FightingKnight h h atk def 0, noAttack) 
+usePotionKnight (FightingKnight h maxH atk def boostDef) = (newKnight, noAttack)
+      where newHealth = max (unHealth h + 10) (unHealth maxH)
+            newKnight = FightingKnight (MkHealth newHealth) maxH atk def boostDef
+
+castDefSpellKnight :: Knight -> (Knight, Attack) 
+castDefSpellKnight (Knight  h atk def) = (FightingKnight h h atk def 3,  noAttack) 
+castDefSpellKnight (FightingKnight h hMax atk def _) = (FightingKnight h hMax atk def 3, noAttack)
+
+
+data Monster = Monster
+    { monsterHealth    :: Health
+    , monsterAttack   :: Attack
+    }  | FledMonster
+    deriving (Show)
+
+fleeMonster :: Monster -> (Monster, Attack)
+fleeMonster _ = (FledMonster, MkAttack 0)
+
+attackMonster :: Monster -> (Monster,Attack)
+attackMonster m = (m, monsterAttack m)
+
+
+class Fighter a where
+    takeDamage :: a -> Attack -> a
+    defeated   :: a -> Bool
+    fled       :: a -> Bool
+
+instance Fighter Knight where
+    takeDamage :: Knight -> Attack -> Knight
+    takeDamage (Knight  h atk def) a =  takeDamage (FightingKnight h h atk def 0)  a
+    takeDamage (FightingKnight  h maxH atk def boostDef) attack = FightingKnight remainingHealth maxH atk def (max (boostDef-1) 0)
+              where reductionFactor = if boostDef > 0 then 2 else 1
+                    effectiveDmg = max (unAttack attack - (unDefense def) * reductionFactor) 0
+                    remainingHealth = MkHealth (max (unHealth h - effectiveDmg) 0)
+
+
+    defeated :: Knight -> Bool
+    defeated (Knight h _ _) = unHealth h <= 0
+    defeated (FightingKnight h _ _ _ _) = unHealth h <= 0
+
+    fled :: Knight -> Bool
+    fled _ = False
+
+
+instance Fighter Monster where
+    takeDamage :: Monster -> Attack -> Monster
+    takeDamage (Monster h a) (MkAttack atk) = Monster remainingHealth a
+                where remainingHealth = MkHealth (max (unHealth h - atk) 0)
+    
+    defeated :: Monster -> Bool
+    defeated m = unHealth (monsterHealth m) <= 0
+
+        
+    fled :: Monster -> Bool
+    fled FledMonster = True
+    fled _ = False
+
+rotateOne :: [a] -> [a]
+rotateOne l = take (length l) (drop 1 (cycle l))
+
+fight :: (Fighter a, Fighter b) =>  a -> [a -> (a, Attack)] -> b -> [b -> (b, Attack)] -> FightResult
+fight fighter1 actions1 fighter2 actions2 = go fighter1 actions1 fighter2 actions2 0
+  where 
+  go :: (Fighter a, Fighter b) =>  a -> [a -> (a, Attack)] -> b -> [b -> (b, Attack)] -> Int -> FightResult
+  go fighter1 actions1 fighter2 actions2 round 
+      | (length actions1) == 0  = FirstFighterDefeated
+      | (length actions2) == 0  = SecondFighterDefeated
```

Comment:
`length` is slow for lists, so if you want to check whether the lists are empty or not, it's better to use `null`.
```suggestion
      | null actions1 = FirstFighterDefeated
      | null actions2 = SecondFighterDefeated
```

But since you need access to list elements anyway, it's even better to use pattern matching on lists directly, something like:

```haskell
case (actions1, actions2) of
--  ([], []) -> ... currently this case is not covered by the implementation ...
    ([], _) -> FirstFighterDefeated
    (_, []) -> SecondFighterDefeated
    (action1 : xs, action2 : yx)
        | fled figher1 ->
        ....
```